Add support for basic hotkeys

Tkael opened the pull request:
Does not yet include any i8n (strings are still hardcoded). I'll take care of that before merging.
Resolves #197.

Tkael committed:
Implement a Hotkey Manager window in the Speech Responder

(does not yet save or load any configuration info)

Tkael committed:
Set up a basic Hotkey Manager

Tkael committed:
Only process key events when the window is active.

Tkael committed:
Refine the Hotkeys window.
Add a `Clear` button and classes to help manage hotkey data.
Improve tests for valid key gestures and feedback for invalid gestures.

Tkael committed:
Revise the HotkeyManager class to store a string name for each hotkey and perform the registration only after the user accepts the hotkeys.

Tkael committed:
Refactor and add saving to the EDDI core configuration file.

Tkael committed:
Additional refactoring

Tkael committed:
Add unit tests.

richardbuckle committed:
Fix threading issue with testing WPF classes

WPF controls must be created in a STA (single-threaded apartment) thread.
We use `STATestClass` instead of `TestClass` to ensure this.
See https://www.meziantou.net/mstest-v2-customize-test-execution.htm - however since that article was written MS has incorporated the solution into the MSTest framework.

richardbuckle commented:
While I understand the allure of hard-coding strings with a view to internationalising them later, trust me, it's a siren song. Having been around this loop more times than I can remember, it has ALWAYS cost me more time to find and fix everything than it would have cost me to use internationalised strings from the get-go.

Let's confine the hard-coded strings to this branch and not merge until we are both happy that we've got them all. Very happy to help with this.

richardbuckle commented on Tests/EddiCoreHotkeyTests.cs:
```diff
@@ -0,0 +1,120 @@
+
+namespace Tests
+{
+    [TestClass, TestCategory( "UnitTests" )]
```
Use `STATestClass` instead of `TestClass` to run all tests within an STA "Single Thread Apartment" thread.

This is necessary because we are instantiating WPF objects that need to be on an STA thread, which in turn is because they are built upon the very primitive Windows COM stuff.

This will make all the unit tests run happily. Obviously, only use this when needed, and when we need it, it is a bit of an indication that we don't have a pure test of our own code. But sometimes, especially when low-level interop is necessary, we have to be pragmatic.

richardbuckle replied:
I found this after finding https://www.meziantou.net/mstest-v2-customize-test-execution.htm
Looks like MS incorporated his advice into a subsequent release.

richardbuckle commented:
We should probably disallow all the nav pad group as I am getting conflicts with the UI when I try to bind them.

By that I mean Ins, Del, Home, End, PgUp, PgDn, and all the arrow keys.

richardbuckle commented:
Opening the config window doesn't load the existing setup and in fact clears it.

richardbuckle commented:
Closing the config window doesn't save, only closing EDDI saves it.

Tkael commented:
Check you `eddi.json` config file and make sure that hotkey assignments are actually being saved?

richardbuckle commented:
Yes I have been monitoring it using Notepad++ in "tail" mode.

richardbuckle committed:
Clear button refreshes hotkeyTextBlock

richardbuckle committed:
Stop RegisterAll() from clobbering the gestures

richardbuckle committed:
Prettify hotkeys window

richardbuckle committed:
Fix hotkey config clobbering all hotkeys

We need to take copies of the name and gesture before unregistering and re-registering.

richardbuckle commented on ConfigService/Configurations/EDDIConfiguration.cs:
```diff
@@ -43,6 +43,11 @@ public class EDDIConfiguration : Config
+        // Hotkeys
+
+        [ JsonProperty( "Hotkeys" ) ]
+        public Dictionary<string, string> Hotkeys { get; set; } = new Dictionary<string, string>();
```
Making a mutable collection member public never ends well, because client code will stick its fingers in and find ways to mutate it without this class knowing anything about it. This then defeats PropertyChanged mechanisms etc, which turns out to be why the config wasn't being written to disk when expected.

Better to keep the collection private and write public methods for adding and removing items. This has the additional benefit that you can change the implementation without affecting client code.

richardbuckle committed:
Encapsulate the config's Hotkeys dict

Exposing a mutable collection as a property never ends well.

Instead, keep the collection private and write methods for Add, Remove and Enumerate.

Fixes cases where hotkey changes were not getting written to disk.

richardbuckle commented:
> Yes I have been monitoring it using Notepad++ in "tail" mode.

Fixed now.

richardbuckle commented:
Playtested and working perfectly.

richardbuckle commented:
Closing to rebase.
